Ignore return, yield and raises checks in abstractmethods

maltevesper opened the pull request:
Ignore some checks in abstract methods to avoid false positives.

    A function is considered abstract if it is decorated with @abstractmethod,
    and considered pure if it has "no implementation".

    Since Python does not have entirely empty functions, all functions which contain
    one of the following (and possibly a docstring) are considered empty

     - ... (Ellipsis)
     - pass
     - raise NotImplementedException
     - return NotImplementedException
     - only a docstring

Closes #54

maltevesper commented:
If you need any edits, further tests etc. please let me know. There are styling bits where I am a bit unsure (i.e. using mypy)

maltevesper commented:
I will have a look at the failing tests tomorrow.
I was a bit to optimistic that the integration commit does not break things, and forgot to rerun pytest :/

maltevesper commented:
I rewrote it, I am not sure about the location of the analysis function (new file), but think it is a bit too much to just cram all that into the function description file.

I just saw that the python < 3.8 checks fail, I will have a quick look.

maltevesper commented:
I believe the 3.6 continuous Integration fail to be caused by a build issue, rather than by my changes. I will now await feedback before I take any further action.

terrencepreilly commented:
Ah, there's some dependency breaking the tests on earlier python versions in `master`.  I haven't had time to look into it.  If it's passing for others, that's probably fine, as I'll look into it before I merge this into `master`.

terrencepreilly commented:
Cool, this looks pretty good so far.  Could you put this into an `ast.NodeVisitor` subclass, and incorporate this into the `AnalysisVisitor`?  The main point of introducing the `AnalysisVisitor` was to have a clean way to incorporate new analysis.

If it's confusing how to add a class to it, I did a minimal example based on your branch in 382e2de8a56a544f62ced97af261b76d00b0808a.  Just make sure to move all of the functionality into the same file, and update the tests to use the visitor.

Thanks for the help!

maltevesper commented:
As you probably saw in fd58822583af0233f3d313a31a6d5573e41528c3 I tried the visitor before and thought that I would break things. I learned a new thing about pythons `super()`, the comment from your commit 👍  did it for me. Thanks for that :).

And I thought a PR would only help me silence the linters, now I actually learned that the mixin pattern is more powerful than I thought if applied correctly.

I noticed that the pattern using super should be rolled out to a wider codebase, I did so in 937a9ef2e404a957d885ec6cf387d1b76984143d. Please read the commit message of that comment for some details and the thought process behind it.

As it turned out I had a flaw in my reasoning which is why I had to do 425077bc3fc48aac4728e35877a9d57e3c1d8cf7. (I think the problem is more related to not calling a visit_* over visit_generic, but I had the patch written with super and thought that I did not want to commit an unnecessary super change)

maltevesper commented:
I added some code deduplication, at this point in time I am not sure if the new function should or should not be a static method. On the one hand, it is a static method (no use of `self`), on the other hand, I would like to keep things consistent.

(There are also conflicting style guides on saying if it is not using self make it static, others saying if it is a static method just turn it into a free function).

Another bit that needs review (apart from the whole thing ;) ) is the use of functools.wraps while it is a good tool, I think it has little purpose where I added it. There are no docstrings to copy in the code base and the parameters are already tied down.

maltevesper commented:
I snuck in a minor change to reindent. Not sure if it is better to suggest it here or in a separate PR (although it is a tiny change). If you feel it should go somewhere else/is a bad change you could just cherry-pick the rest.

terrencepreilly commented on darglint/analysis/visitor_base.py:
```diff
@@ -0,0 +1,27 @@
+
+
+class VisitorBase(ast.NodeVisitor):
+    def continue_visiting(function):
```
There's a lot of weird stuff going on here.  As far as I can tell, this just calls the `super`'s method with the same name, or generic visit.  It's confusing for it to be in a decorator.  It's immediately obvious what

```
    ...
    super().visit_FunctionDef(node)
```

does to anyone reading it in the future.  It's not obvious what

```
    @VisitorBase.continue_visiting
    def visit_FunctionDef(self, node):
        ...
```

does, because it'll do something different depending on its class hierarchy.  The first syntax suggests that it's supposed to have a superclass with a `visit_FunctionDef` method, but the second doesn't.

Let's get rid of it.  I also don't think it's necessary to check if the super function exists -- it'll raise an exception if it doesn't exist.  If you want to be clear about the intent, then just document that a parent must have defined `visit_FunctionDef`, or whatever method you're in.

maltevesper replied:
First of all, awesome that you manage this project and keep it going besides your private life. I do not think you are slow at all :).

Second of all, I am afraid this PR is going to get messy, because there is something complicated at play here: We really need `getattr(super(), f"visit_{node}", super().generic_visit)(node)`.
I only figured it out because I stumbled over it with the automatic test suite (yay for ci). To see why this line is needed I will say only the following before giving you a toy example:

The case of Interest is that we have two `ast` mixins that define the same `visit_XXX` method. Note that the `ast` base does not define `visit_XXX`. With that here is a minimal working example:
```
class MixinDummy:
   """"This is the ast class reduced to the important bits""
    def visit(self, node):
        print(f"MixinDummy visits {node}")
        getattr(self, f"visit_{node}", self.generic_visit)(node)

    def generic_visit(self, node):
        print(
            f"Mixin generic_visit {node}. We end it all now."
            "  Actually ast calls visit on the children, which means:"
            "  we do not see duplicate visit functions for node anymore."
            )

class MixinA(MixinDummy):
   """A visitor in darglint"""
    def visit_a(self, node):
        print("MixinA visit_a {node}")
        #super().generic_visit(node)    # falls short: after visiting MixinA, it would only visit A's children
        #super().visit(node)            # endless recursion, for some reason super().visit always finds this visit_a again...: visit -> visit_a -> visit
        #super().visit_a(node)          # fails if super has no visit_a..., which our mixin dummy (or the ast.NodeVisitor does not)
                                       # With knowledge about other mixins in Mixer and their order, it is possible to decide if we need super.visit_a or
                                       # super.generic_visit here
        getattr(super(), "visit_a", super().generic_visit)(node) # this just works :) (without knowing about other mixins)

class MixinB(MixinDummy):
   """Another visitor in darglint"""
    def visit_a(self, node):
        print("MixinB visit_a {node}")
        #super().generic_visit(node)
        #super().visit(node)
        #super().visit_a(node)
        getattr(super(), "visit_a", super().generic_visit)(node)

class Mixer(MixinA, MixinB):
    """darglint's AnalysisVisitor"""
    pass

Mixer().visit("a")
```

If you play around with this (the different super calls at the end) you will find that the slightly convoluted line `getattr(super(), "visit_a", super().generic_visit)(node)` is needed if you want to maintain all visitors independently of each other.

Furthermore, every visitor **must** call this line, to avoid stoping other visitors short. Currently, the raises visitor is broken in this regard, since the `visit_try` function calls `visit` on child nodes and manipulates the traversal order. Other Mixins do not get the traversal that they might expect below the `try` ast node. Furthermore, other mixins `visit_try` methods might not be called due to this breach of contract.

This does not matter at the moment, since there is no other visitor defining a `visit_Try`.

I had another look at the raise_visitor and will patch the easy half of it, and leave a comment in the bad half.
Once you have digested this flood of information (for me it was helpful to read up on the [MRO](https://www.python.org/download/releases/2.3/mro/) again (link might look old but [the 3.9 docs link there too](https://docs.python.org/3/reference/datamodel.html#the-standard-type-hierarchy) when searching for MRO)), the question is how do we move forward:

I completely understand that you find a class based decorator confusing (took me a bit to figure out how to do that, and it needs to be class based for super to work) and I am happy to reinsert
`return getattr(super(), "visit_X", super().generic_visit)(node)` calls into all methods.

- [ ] Let me know if the decorator should be removed (I think your answer here will be yes).
- [ ] Let me know if you prefer `function.__name__` in the return lines to keep it generic, or if you prefer to have the literal `visit_x` 
- [ ] Let me know what we should do with the new base class: it will only contain one helper function, I could revert the visitors back to inherit `ast` again, but would need to know where the best place for the helper function `_has_decorator` is.

Sorry about the Task list, but I found with long mails/posts like this things get lost. (And I am to lazy to edit the files 3 times if doing it once right is enough ;) )

If you have a look at the other pull request #161 that should be a lot easier to merge :) (unless there is a need to stick to the old ubuntu).

terrencepreilly replied:
> If you play around with this (the different super calls at the end) you will find that the slightly convoluted line getattr(super(), "visit_a", super().generic_visit)(node) is needed if you want to maintain all visitors independently of each other.

They're mixins, so it's not really important that they be independent of one another.  They're primarily just a way of organizing the code and of adding optional functionality.  So we can get rid of checking for the existence of the super method.  Which means the decorator and `VisitorBase` are also not needed.

terrencepreilly reviewed (changes requested):
Sorry if I'm slow to respond to reviews, etc.  I have little time for side projects, since I have to take care of my son all day after work.

maltevesper commented:
I feel apart from the `function_scoped_visitor` mixin I have applied the changes requested in the first review round.
If I would revert the `funciton_scoped_visitor` changes I run into the issue described above:

`function_scoped_visitor` and `abstract_callable_visitor` both define the same `visit_*` functions (i.e. `visit_FunctionDef`). Without the changes in `function_scoped_visitor` mixin , `abstact_callable_visitor.visit_FunctionDef` would not be called.

terrencepreilly commented:
Thanks for the PR! I'll try to include it in a release this weekend.

maltevesper commented:
Thanks for sticking with it. A new release of darglint would be nice.
